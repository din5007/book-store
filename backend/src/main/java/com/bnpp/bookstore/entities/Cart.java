package com.bnpp.bookstore.entities;

import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(
  name = "cart"
  //  uniqueConstraints = { @UniqueConstraint(columnNames = { "userName", "book_id" }) }
)
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
public class Cart {
  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  private Long id;

  @ManyToOne
  @JoinColumn(name = "user_name")
  private User user;

  @ManyToOne
  @JoinColumn(name = "book_id")
  private Book book;

  private int quantity;
}
