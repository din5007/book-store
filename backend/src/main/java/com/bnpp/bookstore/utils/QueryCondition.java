package com.bnpp.bookstore.utils;

import io.micrometer.common.util.StringUtils;
import org.springframework.stereotype.Component;

@Component("queryUtil")
public class QueryCondition {

  public boolean checkNull(String field) {
    return StringUtils.isBlank(field);
  }
}
