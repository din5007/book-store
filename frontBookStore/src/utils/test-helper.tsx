import { StoreProvider } from "easy-peasy";
import { render } from '@testing-library/react';
import { BrowserRouter } from "react-router-dom";


function customRender(component : any , store?: any) {

    return render(<StoreProvider store={store}>
        <BrowserRouter>
            {component}
        </BrowserRouter>
    </StoreProvider>);
}

export * from '@testing-library/react';
export { customRender } ;