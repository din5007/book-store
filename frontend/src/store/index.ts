import { HttpService } from "../service/http/http.service";
import { createStore } from 'easy-peasy';
import { BookModel, booksModel } from "./books/books.model";
import { cartModel, CartModel } from "./cart/cart.model";
import { userModel, UserModel } from "./users/user.model";
import { OrderModel, orderModel } from "./order/order.model";

export interface AppStoreModel {
    books: BookModel;
    cart: CartModel;
    user: UserModel;
    order: OrderModel;
}

export const appStoreModel : AppStoreModel = {
    books: booksModel,
    cart: cartModel,
    user: userModel,
    order: orderModel,
}

export function storeModel() {
    return createStore(appStoreModel, {
        injections: {
            httpService: new HttpService()
        },
        initialState: {
            
        },
    })
}
export const store =  storeModel();