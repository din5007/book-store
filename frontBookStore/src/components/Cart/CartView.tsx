import { Row, Col } from 'react-bootstrap';
import Button from 'react-bootstrap/Button';
import Card from 'react-bootstrap/Card';
import { useStoreActions, useStoreState } from '../../store/hooks';
import { useEffect } from 'react';
import Counter from '../Counter/Counter';

function CartViewList() {
  const { fetchUserCart } = useStoreActions((actions) => actions.cart);
  const { cart } = useStoreState((state) => state.cart);

  useEffect(() => {
    fetchUserCart({});
  }, []);

  return (
    <Row xs={1} md={2} lg={4} className="g-4 m-auto">
        {cart?.map(({book, quantity}, idx) => 
            <Col  key={idx}>
                <Card style={{ width: '18rem' }}>
                <Card.Body>
                    <Card.Title>{book?.title }</Card.Title>
                    <Card.Text>{book?.author}</Card.Text>
                </Card.Body>
                <Card.Footer className='d-flex align-items-baseline justify-content-center text-center'>
                  <Counter id={book?.id} quantity={quantity}/>
                </Card.Footer>
                <Button type="button" size="sm" variant='outline-warning' onClick={() => {}}>Remove from the cart</Button> 
                </Card>
            </Col>
        )}
    </Row>
  );
}

export { CartViewList };