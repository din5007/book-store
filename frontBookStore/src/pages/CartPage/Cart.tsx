import { Button, Container, Row } from "react-bootstrap"
import CartViewList from "../../components/Cart"
import { FC } from "react"
import { useStoreState } from "../../store/hooks"
import { useNavigate } from "react-router"

const Cart : FC<{}> = () => {
    const navigate = useNavigate();
    const { cart } = useStoreState((state) => state.cart);

    return (<div style={{ margin: '5%'}}>
      <Container>
        <Row>
          <div className="page d-flex justify-content-between">
            <h1>You Cart View</h1>
            <Button variant="primary" size="lg" disabled={!cart.length} onClick={() => navigate('/order-summary')}> Go to Summary !</Button> 
          </div>
          <CartViewList />
        </Row>
      </Container>
    </div>)
}

export default Cart;