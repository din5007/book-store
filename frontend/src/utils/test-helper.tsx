import { StoreProvider } from "easy-peasy";
import { render } from '@testing-library/react';
import { AppStoreModel, store } from "../store";
import { Router } from 'react-router-dom';
import { ReactNode } from "react";



// const provider : React.FC<{children : ReactNode, store: any}> = ({ children, store }) => {
//     return (
        
//     )
// }

function customRender(component : any , store: any) {

    return render(<StoreProvider store={store}>
        {component}
    </StoreProvider>);
}

export * from '@testing-library/react';
export { customRender } ;