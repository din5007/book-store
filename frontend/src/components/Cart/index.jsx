import { CartViewList } from "./CartView";

export default CartViewList;