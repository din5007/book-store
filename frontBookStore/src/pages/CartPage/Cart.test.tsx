import { createStore } from 'easy-peasy';
import { customRender as render, waitFor } from '../../utils/test-helper';
import Cart from "./Cart"
import { appStoreModel } from '../../store';
import { describe, expect, it, vi } from 'vitest';

const mock = {
    get: vi.fn()
};

const store = createStore(appStoreModel, {
    injections: {
        httpService : mock
    },
    initialState: {
        cart : {
            cart: [{ book: { id : 1, title : 'Harry Potter', author: 'JK rowling', price: 400, quantity: 400}, quantity: 400, user: {}}]
        }
    }
})

describe('Cart Page', () => {
    it('Cart Page should load properly - snapshot', async () => {
        const {container}  = render(<Cart/>, store);
        waitFor(() => {
            expect(container).toMatchSnapshot();
        })
    });

    it('Cart Page should contain Search component', async () => {
        const {getByText}  = render(<Cart/>, store);
        waitFor(() => {
            expect(getByText('You Cart View')).toBeDefined();
        })
    });

    it('Cart Page should contain Button component', async () => {
        const {getByRole}  = render(<Cart/>, store);
        waitFor(() => {
            expect(getByRole('button', { name : 'Go to Summary'})).toBeDefined();
        })
    });
})