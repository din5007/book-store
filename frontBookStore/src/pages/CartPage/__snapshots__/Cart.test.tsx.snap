// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Page > Cart Page should load properly - snapshot 1`] = `
<div>
  <div
    class="mainback"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="page d-flex justify-content-between"
        >
          <h1>
            You Cart View
          </h1>
          <button
            class="btn btn-primary btn-lg"
            type="button"
          >
             Go to Summary !
          </button>
        </div>
        <div
          class="g-4 m-auto row row-cols-lg-4 row-cols-md-2 row-cols-1"
        >
          <div
            class="col"
          >
            <div
              class="card"
              style="width: 18rem;"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title h5"
                >
                  Harry Potter
                </div>
                <p
                  class="card-text"
                >
                  JK rowling
                </p>
              </div>
              <div
                class="d-flex align-items-baseline justify-content-center text-center card-footer"
              >
                <button
                  class="btn btn-primary btn-sm"
                  type="button"
                >
                  -
                </button>
                <p
                  class="card-text"
                >
                   
                  400
                   
                </p>
                <button
                  class="btn btn-primary btn-sm"
                  type="button"
                >
                  +
                </button>
              </div>
              <button
                class="btn btn-outline-warning btn-md"
                type="button"
              >
                Remove from the cart
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
